### PR TITLE
React Native Biometrics

### PR DESCRIPTION
**Pull request by susie-stytch**
Adds the following example flows to the demo app:

- Log in with biometrics, if a biometric factor already exists
- When logged in:
  - Register biometric factor, if keystore is available
  - Authenticate with biometrics, if a biometric factor already exists
  - Delete biometric factor

| iOS  | Android |
| -- | -- |
| iOS Welcome Page | Android Welcome Page |
| iOS Profile Page | Android Profile Page |
| | Android Register Biometrics |

### CODE REVIEW

File: App.tsx [hunk 1]
```diff
@@ -1,9 +1,13 @@
...
+import {ProfilePage, SendOTPPage, VerifyOTPPage, WelcomePage} from './pages';
+import {NavigationContainer} from '@react-navigation/native';
+import {createNativeStackNavigator} from '@react-navigation/native-stack';
 
-import { StytchClient, StytchProvider, useStytchUser } from '@stytch/react-native';
+import {
+  StytchClient,
+  StytchProvider,
+  useStytchUser,
+} from '@stytch/react-native-testing';
```
Comment (spencer-stytch) [inline App.tsx:10]: remember to revert

File: android/app/src/main/java/stytch/rn/test/newarchitecture/MainApplicationReactNativeHost.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
 import com.facebook.react.fabric.FabricJSIModuleProvider;
 import com.facebook.react.fabric.ReactNativeConfig;
 import com.facebook.react.uimanager.ViewManagerRegistry;
-import com.stytchdemo.BuildConfig;
+import stytch.rn.test.BuildConfig;
 import com.stytchdemo.newarchitecture.components.MainComponentsRegistry;
```
Comment (spencer-stytch) [inline android/app/src/main/java/stytch/rn/test/newarchitecture/MainApplicationReactNativeHost.java:23]: Why are these different prefixes?

File: pages/ProfilePage.tsx [hunk 3]
```diff
@@ -1,11 +1,88 @@
-import { Text, TouchableOpacity, View } from "react-native";
-import sharedStyles from "./shared";
-import { useStytch } from "@stytch/react-native";
-import * as React from 'react'
+import {Alert, Text, TouchableOpacity, View} from 'react-native';
+import sharedStyles from './shared';
+import {useStytch} from '@stytch/react-native-testing';
```
Comment (spencer-stytch) [inline pages/ProfilePage.tsx:3]: revert

File: pages/ProfilePage.tsx [hunk 4]
```diff
@@ -1,11 +1,88 @@
...
+        setSensorType(biometryType);
+      })
+      .catch(err => console.log(err));
+  };
+
+  useEffect(() => {
+    checkSensor();
+    checkKeystoreAvailable();
+    checkBiometricRegistration();
+  });
```
Comment (spencer-stytch) [inline pages/ProfilePage.tsx:39]: add dependency array

File: pages/ProfilePage.tsx [hunk 5]
```diff
@@ -1,11 +1,88 @@
...
+
+  const registerBiometrics = async () => {
+    stytch.biometrics
+      .register({
+        prompt: 'Register biometrics',
+        allowDeviceCredentials: true,
+        allowFallbackToCleartext: false,
+        cancelButtonText: 'Cancel reg',
```
Comment (spencer-stytch) [inline pages/ProfilePage.tsx:47]: Maybe just 'Cancel' since this is supposedly user facing

File: pages/ProfilePage.tsx [hunk 6]
```diff
@@ -1,11 +1,88 @@
...
+      });
+  };
+
+  const authenticateBiometrics = () => {
+    stytch.biometrics
+      .authenticate({
+        prompt: 'Authenticate biometrics',
+        sessionDurationMinutes: 60,
+        allowDeviceCredentials: true,
+        cancelButtonText: 'Cancel Auth',
```
Comment (spencer-stytch) [inline pages/ProfilePage.tsx:64]: 'Cancel' is probably OK here

File: pages/ProfilePage.tsx [hunk 7]
```diff
@@ -16,7 +93,41 @@ function ProfilePage() {
           app on Github to learn how to implement this yourself!
         </Text>
       </View>
-      <TouchableOpacity style={[sharedStyles.buttonDark]} onPress={stytch.session.revoke}>
+      <View>
+        <Text>Biometry type available: {sensorType}</Text>
+      </View>
+      {isKeystoreAvailable && (
```
Comment (spencer-stytch) [inline pages/ProfilePage.tsx:99]: `isKeystoreAvailable && hasBiometricRegistration && (`

File: pages/ProfilePage.tsx [hunk 8]
```diff
@@ -16,7 +93,41 @@ function ProfilePage() {
...
+          <TouchableOpacity
+            style={sharedStyles.buttonDark}
+            onPress={authenticateBiometrics}>
+            <Text style={sharedStyles.buttonTextDark}>
+              Authenticate with Biometrics
+            </Text>
+          </TouchableOpacity>
+        </View>
+      )}
+      <View>
```
Comment (spencer-stytch) [inline pages/ProfilePage.tsx:119]: `{hasBiometricRegistration && (

File: pages/VerifyOTPPage.tsx [hunk 9]
```diff
@@ -9,57 +9,56 @@ import {
...
-} from "react-native";
-import { NativeStackScreenProps } from "@react-navigation/native-stack";
-import { RootStackParamList } from "../App";
-import sharedStyles from "./shared";
-import { useStytch } from '@stytch/react-native'
+} from 'react-native';
+import {NativeStackScreenProps} from '@react-navigation/native-stack';
+import {RootStackParamList} from '../App';
+import sharedStyles from './shared';
+import {useStytch} from '@stytch/react-native-testing';
```
Comment (spencer-stytch) [inline pages/VerifyOTPPage.tsx:16]: revert

File: pages/WelcomePage.tsx [hunk 10]
```diff
@@ -1,24 +1,57 @@
-import { Image, Text, TouchableOpacity, View } from "react-native";
-import { NativeStackScreenProps } from "@react-navigation/native-stack";
-import { RootStackParamList } from "../App";
-import styles from "./shared";
-import * as React from 'react'
+import {Alert, Image, Text, TouchableOpacity, View} from 'react-native';
+import {NativeStackScreenProps} from '@react-navigation/native-stack';
+import {RootStackParamList} from '../App';
+import {useStytch} from '@stytch/react-native-testing';
```
Comment (spencer-stytch) [inline pages/WelcomePage.tsx:4]: revert

File: pages/WelcomePage.tsx [hunk 11]
```diff
@@ -1,24 +1,57 @@
...
+
+  const checkBiometricsRegistration = async () => {
+    await stytch.biometrics.isRegistrationAvailable().then(resp => {
+      setBiometricRegistration(resp);
+    });
+  };
+
+  useEffect(() => {
+    checkBiometricsRegistration();
+  });
```
Comment (spencer-stytch) [inline pages/WelcomePage.tsx:23]: add dependency array

**Review by spencer-stytch — APPROVED** [review]
Nicely done! Again, just a couple small tweaks and remember to revert the package name, but otherwise g2g!

File: android/app/src/main/java/stytch/rn/test/newarchitecture/MainApplicationReactNativeHost.java (hunk 2 above)
Comment (susie-stytch) [reply android/app/src/main/java/stytch/rn/test/newarchitecture/MainApplicationReactNativeHost.java:23]: I found that stytch.rn.test is where the BuildConfig is generated after building the app, and com.stytchdemo doesn't contain it. Not too sure why this is, I'll try to investigate if there's something we can do to change it

**Pull request merged**